state/pruner.go

```diff
@@ -274,6 +350,75 @@ func (p *Pruner) pruneBlocksRoutine() {
 	}
 }
 
+func (p *Pruner) pruneIndexesRoutine() {
+	p.logger.Info("Index pruner started", "interval", p.interval.String())
+	lastTxIndexerRetainHeight := int64(0)
+	lastBlockIndexerRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			lastTxIndexerRetainHeight = p.pruneTxIndexerToRetainHeight(lastTxIndexerRetainHeight)
+			lastBlockIndexerRetainHeight = p.pruneBlockIndexerToRetainHeight(lastBlockIndexerRetainHeight)
+			// TODO call observer
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneTxIndexerToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight, err := p.GetTxIndexerRetainHeight()
+	if err != nil {
+		// Indexer retain height has not yet been set - do not log any
+		// errors at this time.
+		if errors.Is(err, ErrKeyNotFound) {
+			return 0
+		}
+		p.logger.Error("Failed to get Indexer retain height", "err", err)
+		return lastRetainHeight
+	}
+
+	if lastRetainHeight >= targetRetainHeight {
+		return lastRetainHeight
+	}
+
+	numPrunedTxIndexer, newTxIndexerRetainHeight, err := p.txIndexer.Prune(targetRetainHeight)
+	if err != nil {
+		p.logger.Error("Failed to prune tx indexer", "err", err, "targetRetainHeight", targetRetainHeight, "newTxIndexerRetainHeight", newTxIndexerRetainHeight)
+	} else if numPrunedTxIndexer > 0 {
+		// TODO call metrics
+		p.logger.Info("Pruned tx indexer", "count", numPrunedTxIndexer, "newTxIndexerRetainHeight", newTxIndexerRetainHeight)
+	}
+	return newTxIndexerRetainHeight
+}
+
+func (p *Pruner) pruneBlockIndexerToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight, err := p.GetBlockIndexerRetainHeight()
+	if err != nil {
+		// Indexer retain height has not yet been set - do not log any
+		// errors at this time.
+		if errors.Is(err, ErrKeyNotFound) {
+			return 0
+		}
+		p.logger.Error("Failed to get Indexer retain height", "err", err)
+		return lastRetainHeight
+	}
+
+	if lastRetainHeight >= targetRetainHeight {
+		return lastRetainHeight
+	}
+
+	numPrunedBlockIndexer, newBlockIndexerRetainHeight, err := p.blockIndexer.Prune(targetRetainHeight)
+	if err != nil {
+		p.logger.Error("Failed to prune block indexer", "err", err, "targetRetainHeight", targetRetainHeight, "newBlockIndexerRetainHeight", newBlockIndexerRetainHeight)
+	} else if numPrunedBlockIndexer > 0 {
+		// TODO call metrics
+		p.logger.Info("Pruned block indexer", "count", numPrunedBlockIndexer, "newBlockIndexerRetainHeight", newBlockIndexerRetainHeight)
```

Comment:
```suggestion
		p.logger.Debug("Pruned block indexer", "count", numPrunedBlockIndexer, "newBlockIndexerRetainHeight", newBlockIndexerRetainHeight)
```

We will introduce metrics to monitor this.